src/torcontrol.cpp

```diff
@@ -0,0 +1,673 @@
+    struct event *reconnect_ev;
+    float reconnect_timeout;
+    CService service;
+    /** Cooie for SAFECOOKIE auth */
+    std::vector<uint8_t> cookie;
+    /** ClientNonce for SAFECOOKIE auth */
+    std::vector<uint8_t> clientNonce;
+
+    /** Callback for ADD_ONION result */
+    void add_onion_cb(TorControlConnection& conn, const TorControlReply& reply);
+    /** Callback for AUTHENTICATE result */
+    void auth_cb(TorControlConnection& conn, const TorControlReply& reply);
+    /** Callback for AUTHCHALLENGE result */
+    void authchallenge_cb(TorControlConnection& conn, const TorControlReply& reply);
+    /** Callback for PROTOCOLINFO result */
+    void protocolinfo_cb(TorControlConnection& conn, const TorControlReply& reply);
+    /** Callback after succesful connection */
+    void connected_cb(TorControlConnection& conn);
+    /** Callback after connection lost or failed connection attempt */
+    void disconnected_cb(TorControlConnection& conn);
+
+    /** Callback for reconnect timer */
+    static void reconnect_cb(evutil_socket_t fd, short what, void *arg);
+};
+
+TorController::TorController(struct event_base* base, const std::string& target):
+    base(base),
+    target(target), conn(base), reconnect(true), reconnect_ev(0),
+    reconnect_timeout(RECONNECT_TIMEOUT_START)
+{
+    // Start connection attempts immediately
+    if (!conn.Connect(target, boost::bind(&TorController::connected_cb, this, _1),
+         boost::bind(&TorController::disconnected_cb, this, _1) )) {
+        LogPrintf("tor: Initiating connection to Tor control port %s failed\n", target);
+    }
+    // Read service private key if cached
+    std::pair<bool,std::string> pkf = ReadBinaryFile(GetPrivateKeyFile());
+    if (pkf.first) {
+        LogPrint("tor", "tor: Reading cached private key from %s\n", GetPrivateKeyFile());
+        private_key = pkf.second;
+    }
+}
+
+TorController::~TorController()
+{
+    if (reconnect_ev)
+        event_del(reconnect_ev);
+    if (service.IsValid()) {
+        RemoveLocal(service);
+    }
+}
+
+void TorController::add_onion_cb(TorControlConnection& conn, const TorControlReply& reply)
+{
+    if (reply.code == 250) {
+        LogPrint("tor", "tor: ADD_ONION succesful\n");
+        BOOST_FOREACH(const std::string &s, reply.lines) {
+            std::map<std::string,std::string> m = ParseTorReplyMapping(s);
+            std::map<std::string,std::string>::iterator i;
+            if ((i = m.find("ServiceID")) != m.end())
+                service_id = i->second;
+            if ((i = m.find("PrivateKey")) != m.end())
+                private_key = i->second;
+        }
+
+        service = CService(service_id+".onion", GetListenPort(), false);
+        LogPrintf("tor: Got service ID %s, advertizing service %s\n", service_id, service.ToString());
+        if (WriteBinaryFile(GetPrivateKeyFile(), private_key)) {
+            LogPrint("tor", "tor: Cached service private key to %s\n", GetPrivateKeyFile());
```

Comment:
Also, mind to pass this file through

`contrib/devtools/clang-format.py clang-format src/torcontrol.cpp `?